src/SourcePathsMappingUI/AskUserForFile.cpp

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "SourcePathsMappingUI/AskUserForFile.h"
+
+#include <QAbstractButton>
+#include <QCheckBox>
+#include <QDir>
+#include <QFileDialog>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QSettings>
+#include <memory>
+
+#include "OrbitBase/ReadFileToString.h"
+#include "SourcePathsMapping/Mapping.h"
+#include "SourcePathsMapping/MappingManager.h"
+
+namespace orbit_source_paths_mapping_ui {
+
+std::optional<QString> TryAskingTheUserAndReadSourceFile(QWidget* parent,
+                                                         const std::filesystem::path& file_path) {
+  QMessageBox message_box{QMessageBox::Warning, "Source code file not found",
+                          QString("Could not find the source code file \"%1\" on this machine.")
+                              .arg(QString::fromStdString(file_path.string())),
+                          QMessageBox::Cancel, parent};
+  auto* pick_file_button = message_box.addButton("Choose file...", QMessageBox::ActionRole);
```

Comment:
I know that this was probably there before and should have been spotted in another PR, but most (all?) of the `auto` declarations should be replaced with types.

Reply:
done.